DM-15204: Chained datastores

timj opened the pull request:
Simple chained datastore supporting multiple datastores where a put is sent to all and a get is retrieved from the first item.

timj committed:
Fix sphinx build of documentation

timj committed:
Augment dummy registry to use datastore name

This is critical for chained datastores

timj committed:
First attempt at ChainedDatastore

timj committed:
Add support for lists in Config

YAML supports lists and we want to be able to represent
those lists in the Config class. This mostly works
although Config will not automatically grow a list if a higher
index is specified.

timj committed:
Add config helper method for updating entries

timj committed:
Update setConfigRoot to use shared helper code

timj committed:
Support default expansion for configs that contain other configs

Such as chained datastores.

timj committed:
Add setConfigRoot for chained datastores

timj committed:
Add chained datastore tests to butler tests

timj committed:
Add __contains__ to StorageClassFactory

This makes it easy to check whether a storage class has already
been registered.

timj committed:
Compare storage classes when registering

This brings the code in to line with the documentation although
the equality test still only compares names.

timj committed:
Fix order bug in adding storage classes from config

The config defining storage classes does not guarantee
to return definitions in the order foudn in the YAML file
so this change ensures that all the non-composite
storage classes are defined before the composite
storage classes that use those others are defined.

timj committed:
Use full config to determine if this is a parent config or not

timj committed:
Remove code in set item that does double assignment

It may be that I don't understand what is happening here
but the code being removed sets the value using `update`
but is immediately followed by code that does almost the
same thing and assigns the new value a second time.

Removing this code causes no tests to fail.

timj committed:
Handle config assignment to config

Deep copy the supplied config (as happens for constructors)
if a Config is supplied to __setitem__

timj committed:
Reorganize to reduce code duplication and enable chained datastores

Uses setConfigRoot to allow the code to be more generic.

timj committed:
Add an implementation of ingest for chained datastores

timj committed:
Update Butler.makeRepo test to work with all datastores

timj commented:
@TallJimbo I think it's worth you having a look at the current implementation now. The tests now work with a ChainedDatastore and I've had to rewrite the config system a bit and overhaul setConfigRoot to get this to all work. StorageClassFactory had a bit of an overhaul as well to handle the order of definitions not being guaranteed (I am leaving the equality test for new definitions to the other ticket).

TallJimbo commented on python/lsst/daf/butler/core/config.py:
```diff
@@ -406,6 +426,80 @@ def dumpToFile(self, path):
             self.dump(f)
 
+    @staticmethod
+    def overrideConfigParameters(configType, config, full, toupdate=None, tocopy=None):
```
Given that the class is already called `Config`, including "Config" in the name of the method seems redundant.

`toupdate`, `tocopy` -> camelCase.

TallJimbo commented on python/lsst/daf/butler/core/datastore.py:
```diff
@@ -142,6 +142,11 @@ class Datastore(metaclass=ABCMeta):
+    containerKey = None
+    """Name of the key containing a list of subconfigurations that also
+    need to be merged with defaults and may use different Python types
+    (but the same config class). Assumed to be a list of configs."""
```
Does this actually need to be here?  Seems like it's something only some subclasses need.  Or was this supposed to be part of `DatastoreConfig` rather than `Datastore`?

timj replied:
`ConfigSubset` needs to know the container key used by each class, so it's definitely a datastore property and not a config property: we only have one DatastoreConfig but each datastore class has a different content in those configs: we could have DatastoreConfig subclasses but we don't do it that way at the moment). Saying that, I can remove this from here because `ConfigSubset` traps the case where the value hasn't been set for the class -- it was here mainly as a documentation aid for the property.

TallJimbo commented on python/lsst/daf/butler/core/sqlDatabaseDict.py:
```diff
@@ -55,12 +57,13 @@ class SqlDatabaseDict(DatabaseDict):
+        `~collections.namedtuple`); these field names must also appear
         in the ``types`` arg, and a `_make` attribute to construct it from a
-        sequence of values (again, as defined by `namedtuple`).
+        sequence of values (again, as defined by `~collections.namedtuple`).
```
Just for my own education: what does the `~` do here?

timj replied:
It writes the documentation as `namedtuple` but tries to link to `collections.namedtuple`.

TallJimbo commented on python/lsst/daf/butler/core/utils.py:
```diff
@@ -24,10 +24,10 @@
 __all__ = ("iterable", "allSlots", "slotValuesAreEqual", "slotValuesToHash",
            "getFullTypeName", "doImport", "getInstanceOf", "Singleton",
-           "TopologicalSet", "transactional", "getObjectSize")
+           "TopologicalSet", "TopologicalSetNode", "transactional",
```
Did you add an external use of `TopologicalSetNode` I missed?  I thought it was an implementation detail.

timj replied:
Without this the sphinx build did not work. I'll give it another try.

timj replied:
Trying again, when I build the docs I get this:
```
lsst.daf.butler.core.utils.TopologicalSetNode.rst:: WARNING: document isn't included in any toctree```
If I leave out `TopologicalSetNode`. It's only used internally in utils.py so I'm not sure why Sphinx insists it should be listed in `__all__`. Maybe @jonathansick has an idea.

TallJimbo commented on python/lsst/daf/butler/datastores/chainedDatastore.py:
```diff
@@ -0,0 +1,342 @@
+
+        # Name ourselves with the timestamp the datastore
+        # was created.
+        self.name = "{}@{}".format(type(self).__qualname__, time.time())
```
If this is timestamp, it's probably not useful as a publicly-visible name; should this be a private attribute?

timj replied:
The name has to be unique (the name is used in registry) so this makes it unique. I can remove this if the system ensures that only one inMemoryDatastore can ever be created in memory.

TallJimbo replied:
Will this go away when Registry supports ephemeral datastores directly, then?

Also, why does a constraint on `InMemoryDatastore` affect `ChainedDatastore`?

timj replied:
Re InMemoryDatastore: Because I got my wires crossed (InMemoryDatastore does need the timestamp to make it unique). You are right that this name is not used directly in registry because the puts are done by the contained datastores (which need unique names). I could just call this `ChainedDatastore` or else call it `ChainedDatastore[datastore1.name, datastore2.name]`

TallJimbo replied:
Either is fine with me - you probably have a better sense than I do whether we want a more unique and descriptive name here or something more concise.

TallJimbo commented on python/lsst/daf/butler/datastores/chainedDatastore.py:
```diff
@@ -0,0 +1,342 @@
+        """
+
+        for datastore in self.datastores:
+            datastore.put(inMemoryDataset, ref)
```
What does a `Datastore` normally do when it isn't configured to support a particular `DatasetRef` when it's asked to `put` (e.g. a `PosixDatastore` that doesn't have a formatter configured)?

If the answer is "raise an exception", we probably either want to change the behavior of `ChainedDatastore` itself to ignore those failures, or add a new trivial forwarding `Datastore` in between that does.  I think it'll be pretty common to want to write only some `DatasetTypes` to certain `Datastores` in the chain.

timj replied:
I agree that if a datastore put fails in a specific way, we can skip that datastore and allow the chained put to work if at least one put worked (possibly to a non-ephemeral datastore -- we could make that configurable).

TallJimbo commented on python/lsst/daf/butler/datastores/chainedDatastore.py:
```diff
@@ -0,0 +1,342 @@
+        them.
+
+        A transfer mode of None is not supported since that requires the
+        file to have been previously copied to each individual datastore.
```
Seems like `None` ought to be okay if only one `Datastore` in the chain supports ingest, and the file is already in that one.  I think you'll get the right behavior if you just remove that special check.

timj replied:
My real problem here was to understand how a file turned up in one of the data stores in the first place. The test infrastructure doesn't actually know that a chained datastore is being used so I wasn't sure how much special casing I should be including to copy that file in to the right place during the tests (it would have to know to copy the test file to `.datastores[1].root`). I now realize that I should add a second posix datastore to the chaining tests since that would demonstrate that ingest with transfer=move would fail on the second datastore. Then we have to decide whether a single ingest working is good enough or if all implemented ingests should work (transfer=move could work if the chained ingest copied the file itself to a temporary area so that each datastore could move it itself)

TallJimbo replied:
What about making it so `ingest` is only supported if exactly one `Datastore` accepts the file?  I can't think of a good use case for anything more complex than that (though we would probably want to let different `Datastores` in the chain accept different files).

timj replied:
Hmm. It seems a bit odd to me that ingest would stop at the first success (or continue trying in order to ensure that they all subsequently failed) but `put` sends to each child datastore.

TallJimbo commented on python/lsst/daf/butler/datastores/chainedDatastore.py:
```diff
@@ -0,0 +1,342 @@
+            Attempt to remove a dataset that does not exist.
+        """
+        for datastore in self.datastores:
+            datastore.remove(ref)
```
Given that `get` works if only one `Datastore` has the dataset, this should probably behave like `ingest` and raise only if it wasn't removed anywhere.

timj replied:
Oh yes. That shows that my tests need to be extended.

TallJimbo reviewed (approved):
Looks good!  A few comments, but nothing big or structural.

timj committed:
Use shorter name for config param override

Since the method is already in Config do not include Config in
the name and use overrideParameters instead. Also switch from
toupdate/tocopy to toUpdate/toCopy.

timj committed:
Improve containerKey documentation

Not strictly necessary to define it at all but it is not
unreasonable to document why it is needed.

timj committed:
Derive ChainedDatastore name from child datastore names

timj committed:
Update docstring and switch TypeError to KeyError

KeyError is more consistent with "item requested is not known
to this object".

timj committed:
Define DatasetTypeNotSupportedError

timj committed:
Use DatasetTypeUnsupportedError and test for it

timj committed:
Rewrite ChainedDatastore to allow failures in child put/ingest/remove

Not all children will support a specific operation.

timj committed:
Add some logging to datastores

timj committed:
Change behavior of ChainedDatastore with ephemeral datastores

* getUri prefers concrete datastores.
* put warns if no concrete datastore accepted the dataset

timj committed:
The 2nd posix datastore must use a different table name

timj commented:
@TallJimbo do you want to take another look? ChainedDatastore tries to be a bit more robust against child datastores not handling certain datasets. Still no per-datasettype configuration on in-memory or chained datastore handling or cache expiry (which we should discuss punting to a different ticket).

timj committed:
Add dynamic ephemeral test to ChainedDatastore
